### PR TITLE
MT Mode flag MOD-4954

### PR DESCRIPTION
**Pull request by DvirDukhan**
This PR modifies the MT modes flags - cosmetics
CC @LiorKogan

### CODE REVIEW

File: build/docker/dockerfile.tmpl [hunk 1]
```diff
@@ -65,8 +65,8 @@ RUN chown -R redis:redis /var/opt/redislabs
 COPY --from=builder /build/bin/artifacts/ /var/opt/redislabs/artifacts/
 {% endif %}
 
-{% if REDISEARCH_POWER_TO_THE_WORKERS == "1" %}
-CMD ["sh", "-c", "redis-server --loadmodule {{MODULES_DIR}}/redisearch.so ALWAYS_USE_THREADS TRUE WORKER_THREADS $(nproc) --loadmodule {{MODULES_DIR}}/rejson.so"]
+{% if REDISEARCH_MT_BUILD == "1" %}
+CMD ["sh", "-c", "redis-server --loadmodule {{MODULES_DIR}}/redisearch.so MT_MODE MT_MODE_RCP WORKER_THREADS $(nproc) --loadmodule {{MODULES_DIR}}/rejson.so"]
```
Comment (rafie) [inline build/docker/dockerfile.tmpl:69]: `MT_MODE_RCP`? For the general public?

File: src/config.h [hunk 2]
```diff
@@ -224,6 +234,14 @@ void DialectsGlobalStats_AddToInfo(RedisModuleInfoCtx *ctx);
 #define SET_DIALECT(barr, d) (barr |= DIALECT_OFFSET(d))     // set the d'th dialect in the dialect bitarray to true.
 #define VECSIM_DEFAULT_BLOCK_SIZE   1024
 
+#ifdef MT_BUILD  
+#define MT_BUILD_CONFIG() .numWorkerThreads = 0,                                                                     \
+    .mt_mode = MT_MODE_OSS,                                                                                                     \
+    .tieredVecSimIndexBufferLimit = DEFAULT_BLOCK_SIZE,                                                               
+#else 
+#define MT_BUILD_CONFIG()
```
Comment (rafie) [inline src/config.h:242]: Can drop the `()`

File: build/docker/dockerfile.tmpl (hunk 1 above)
Comment (alonre24) [reply build/docker/dockerfile.tmpl:69]: I agree, this basically changes the previous configuration (`ALWAYS_USE_THREADS TRUE` is equivalent to `MT_MODE_RCE`)

File: src/config.c [hunk 3]
```diff
@@ -245,10 +245,40 @@ CONFIG_GETTER(getWorkThreads) {
...
+  switch (mt_mode) {
+    case MT_MODE_OSS:
+      return "MT_MODE_OSS";
+    case MT_MODE_RCE:
+      return "MT_MODE_RCE";
+    case MT_MODE_RCP:
+      return "MT_MODE_RCP";
+    default:          // LCOV_EXCL_LINE cannot be reached
+      return "huh?";  // LCOV_EXCL_LINE cannot be reached
+  }
```
Comment (alonre24) [inline src/config.c:276]: Didn't understand the comment...

File: src/config.c [hunk 4]
```diff
@@ -688,18 +718,18 @@ RSConfigOptions RSGlobalConfigOptions = {
...
-        {.name = "ALWAYS_USE_THREADS",
+        {.name = "MT_MODE",
          .helpText = "Let ft.search and vector indexing be done in background threads as default if"
-                        "set to TRUE, use workers thread pool for operational needs only otherwise",
-         .setValue = setThreadsEnabled,
-         .getValue = getThreadsEnabled,
+                        "set to MT_MODE_RCP. MT_MODE_RCE use workers thread pool for operational needs only otherwise",
+         .setValue = setMtMode,
+         .getValue = getMtMode,
          .flags = RSCONFIGVAR_F_IMMUTABLE | RSCONFIGVAR_F_FLAG,
```
Comment (alonre24) [inline src/config.c:733]: If in not mistaken `RSCONFIGVAR_F_FLAG` means a boolean flag (which is not the case anymore...) , though I don't see where we use that flag... @GuyAv46 right?

File: src/config.h [hunk 5]
```diff
@@ -104,11 +114,11 @@ typedef struct {
...
   int poolSizeNoAuto;  // Don't auto-detect pool size
 
-// #ifdef POWER_TO_THE_WORKERS
+#ifdef MT_BUILD
   size_t numWorkerThreads;
-  int alwaysUseThreads;
+  MTMode mt_mode;
   size_t tieredVecSimIndexBufferLimit;
-// #endif
+#endif
```
Comment (alonre24) [inline src/config.h:121]: I think that the `#ifdef` was commented out on purpose, and the build error that you have in coordinator is due to this change...

File: src/config.c (hunk 4 above)
Comment (GuyAv46) [reply src/config.c:733]: Yep. should be removed. And indeed nobody uses this flag anywhere today but...

**Review by rafie — DISMISSED** [review]

**Review by rafie — APPROVED** [review]

File: src/aggregate/aggregate.h [hunk 6]
```diff
@@ -75,9 +75,11 @@ typedef enum {
 #define IsWildcard(r) ((r)->ast.root->type == QN_WILDCARD)
 #define HasScorer(r) ((r)->optimizer->scorerType != SCORER_TYPE_NONE)
 
-// Indicates whether a query should run in the background (only when the immutable alwaysUseThreads
-// config is set). This will also guarantee that there is a running thread pool with al least 1 thread.
-#define RunInThread() (RSGlobalConfig.alwaysUseThreads)
+#ifdef MT_BUILD
+// Indicates whether a query should run in the background. This
+// will also guarantee that there is a running thread pool with al least 1 thread.
+#define RunInThread(r) (RSGlobalConfig.mt_mode == MT_MODE_FULL)
```
Comment (GuyAv46) [inline src/aggregate/aggregate.h:81]:
```suggestion
#define RunInThread() (RSGlobalConfig.mt_mode == MT_MODE_FULL)
```

File: src/aggregate/aggregate_exec.c
Comment (GuyAv46) [inline src/aggregate/aggregate_exec.c:1]: there is another `POWER_TO_THE_WORKERS` on this file that needs to be changed

File: tests/pytests/test_async.py [hunk 7]
```diff
@@ -74,9 +74,9 @@ def test_mod4745(env):
 
 
 def test_eval_node_errors_async():
-    if not POWER_TO_THE_WORKERS:
+    if not MT_BUILD:
         raise unittest.SkipTest("Skipping since worker threads are not enabled")
```
Comment (GuyAv46) [inline tests/pytests/test_async.py:78]: we can use `@skip(noWorkers=True)` now

File: tests/pytests/test_multithread.py [hunk 8]
```diff
@@ -10,14 +10,16 @@
 
 @skip(noWorkers=True)
 def testEmptyBuffer():
-    env = Env(moduleArgs='WORKER_THREADS 1 ALWAYS_USE_THREADS TRUE')
+    if not MT_BUILD:
+        raise unittest.SkipTest("Skipping since worker threads are not enabled")
```
Comment (GuyAv46) [inline tests/pytests/test_multithread.py:14]: redundant with `@skip(noWorkers=True)`

File: tests/pytests/test_multithread.py [hunk 9]
```diff
@@ -43,12 +45,16 @@ def CreateAndSearchSortBy(docs_count):
 
 @skip(noWorkers=True)
 def testSimpleBuffer():
+    if not MT_BUILD:
+        raise unittest.SkipTest("Skipping since worker threads are not enabled")
```
Comment (GuyAv46) [inline tests/pytests/test_multithread.py:49]: redundant with `@skip(noWorkers=True)`

File: tests/pytests/test_multithread.py [hunk 10]
```diff
@@ -43,12 +45,16 @@ def CreateAndSearchSortBy(docs_count):
...
     CreateAndSearchSortBy(docs_count = 10)
 
 # In this test we have more than BlockSize docs to buffer, we want to make sure there are no leaks
 # caused by the buffer memory management.
 @skip(noWorkers=True)
 def testMultipleBlocksBuffer():
+    if not MT_BUILD:
+        raise unittest.SkipTest("Skipping since worker threads are not enabled")
```
Comment (GuyAv46) [inline tests/pytests/test_multithread.py:57]: redundant with `@skip(noWorkers=True)`

File: tests/pytests/test_multithread.py [hunk 11]
```diff
@@ -88,7 +94,9 @@ def get_pipeline(profile_res):
 
 @skip(noWorkers=True)
 def test_pipeline():
-    env = Env(moduleArgs='WORKER_THREADS 1 ALWAYS_USE_THREADS TRUE')
+    if not MT_BUILD:
+        raise unittest.SkipTest("Skipping since worker threads are not enabled")
```
Comment (GuyAv46) [inline tests/pytests/test_multithread.py:98]: redundant with `@skip(noWorkers=True)`

File: tests/pytests/test_multithread.py [hunk 12]
```diff
@@ -362,12 +370,26 @@ def apply(paged= False):
     env.assertEqual(get_pipeline(res), expected_pipeline)
 
 
-def test_invalid_config():
-    if not POWER_TO_THE_WORKERS:
+def test_invalid_MT_MODE_FULL_config():
+    if not MT_BUILD:
+        raise unittest.SkipTest("Skipping since worker threads are not enabled")
```
Comment (GuyAv46) [inline tests/pytests/test_multithread.py:375]:
```suggestion
@skip(noWorkers=True)
def test_invalid_MT_MODE_FULL_config():
```

File: tests/pytests/test_multithread.py [hunk 13]
```diff
@@ -362,12 +370,26 @@ def apply(paged= False):
...
+        env.cmd(f"{prefix}ft.config", "get", "WORKER_THREADS")
+        env.assertFalse(True)   # We shouldn't get here
+    except Exception:
+        # Create dummy env to collect exit gracefully.
+        env = Env()
+        pass
+
+def test_invalid_MT_MODE_ONLY_ON_OPERATIONS_config():
+    if not MT_BUILD:
         raise unittest.SkipTest("Skipping since worker threads are not enabled")
```
Comment (GuyAv46) [inline tests/pytests/test_multithread.py:389]:
```suggestion
@skip(noWorkers=True)
def test_invalid_MT_MODE_ONLY_ON_OPERATIONS_config():
```

File: tests/pytests/test_multithread.py
Comment (GuyAv46) [inline tests/pytests/test_multithread.py:1]: consider using `@skip(noWorkers=True)` on worker-depended tests

**Review by GuyAv46 — COMMENTED** [review]
Some rebase suggestions

File: tests/pytests/test_multithread.py [hunk 14]
```diff
@@ -361,13 +365,22 @@ def apply(paged= False):
     #sortby NOT SORTABLE NOCONTENT
     env.assertEqual(get_pipeline(res), expected_pipeline)
 
+def test_invalid_MT_MODE_FULL_config():
+    env = initEnv()
+    try:
+        env = initEnv(moduleArgs='WORKER_THREADS 0 MT_MODE MT_MODE_FULL')
```
Comment (GuyAv46) [inline tests/pytests/test_multithread.py:371]: double env set is not good

File: tests/pytests/test_multithread.py [hunk 15]
```diff
@@ -378,10 +391,10 @@ def test_invalid_config():
 
 
 def test_reload_index_while_indexing():
-    if not POWER_TO_THE_WORKERS or CODE_COVERAGE:
-        raise unittest.SkipTest("Skipping since worker threads are not enabled")
+    if CODE_COVERAGE:
```
Comment (GuyAv46) [inline tests/pytests/test_multithread.py:394]: no pretty skip on `common.py`?

**Pull request merged**